client/signup/config/flows.js

```diff
@@ -238,11 +239,72 @@ const Flows = {
 			flow = filterDesignTypeInFlow( flow );
 		}
 
-		return flow;
+		return Flows.getABTestFilteredFlow( flowName, flow );
 	},
 
 	getFlows() {
 		return flows;
+	},
+
+	/**
+	 * Preload AB Test variations after a certain step has been completed.
+	 *
+	 * This gives the option to set the AB variation as late as possible in the
+	 * signup flow.
+	 *
+	 * Currently only the `main` flow is whitelisted.
+	 *
+	 * @param {String} flowName The current flow
+	 * @param {String} stepName The step that is being completed right now
+	 */
+	preloadABTestVariationsForStep( flowName, stepName ) {
```

Comment:
The usage of `getABTestVariation` in `getABTestFilteredFlow` is intentional.

The flow should not be modified until the AB test variation is assigned to the user. The assignment should happen at the latest possible moment, which means after the previous step is completed and before the next one starts. 

`getFlow` and subsequently `getABTestFilteredFlow` is being called since the very early moments of signup initialization and this will force the AB test variants to be assigned even before the first step. If users are assigned this early, the test data may be diluted by the users who don't even reach the step we're testing. 

On the other hand `preloadABTestVariationsForStep` is called after every step (or rather before loading a step). This way we can set the AB variation in this exact moment and not earlier.

With this way of setting the logic, `getFlow` is strictly only giving out the current flow that should be and it will get modified by explicit calls to update the AB variations in `preloadABTestVariationsForStep`. I look at this more like keeping the logic separate, so `getFlow` doesn't get too complicated and black-magic-y.

Of course the whole thing may not make any sense :) I would love to hear thoughts on the above.